### PR TITLE
Progress report December 2023

### PR DESCRIPTION
**Pull request by goldenx86**
Too long

### CODE REVIEW

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+{{< single-title-imgs
+    "Bet you enjoy this one the most"
+    "./config4.png"
+    "./config5.png"
+    >}}
+
+- A save data manager, allowing to export or import per game saves, along the global option.
+- An option to delete all save data of that particular game.
+- An option to clear the pipeline cache of that particular game.
+- And a Start button, which allows you to select global or custom configuration. Launching from the game list will always load the custom configuration.
```
Comment (FearlessTobi) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:315]: "which allows you to select **a** global or custom configuration" ?

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+If only the software page table is being used, there’s no problem, as the blocks aren’t resulting in calls to `mmap`.
+
+When yuzu uses host-mapped addressing, the emulator propagates all of these mappings into the host address space.
+That in itself wouldn’t normally be a problem, if it wasn’t for the fact that Unreal Engine 4 can allocate hundreds of thousands of small heap blocks, going over the kernel limit with ease, and crashing the entire program.
+
+yuzu is not the only [project](https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html) affected by this completely arbitrary limitation–it has been a complaint for quite a long time.
+
+To work around this limitation, byte[] {{< gh-hovercard "12466" "inserts a layer" >}} between the software page table and the host mapping system, constantly tracking the heap allocations made by programs and automatically recycling some less recently used mappings, which causes some stuttering―but better than a game crash. Collection will reduce the number of mappings to around half of the 65530 limit most systems ship with.
```
Comment (FearlessTobi) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:361]:
```suggestion
To work around this limitation, byte[] {{< gh-hovercard "12466" "inserts a layer" >}} between the software page table and the host mapping system, constantly tracking the heap allocations made by programs and automatically recycling some less recently used mappings, which causes some stuttering―but is better than a game crash. Collection will reduce the number of mappings to around half of the 65530 limit most systems ship with.
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+By {{< gh-hovercard "12292" "adding a udev rule" >}} to grant access to `hidraw` devices, HurricanePootis circumvented this limitation.
+Thank you!
+
+## Miscellaneous changes
+
+### Multiprocess preliminary work
+
+We spoke about preliminary work for multiprocess support. There’s more to cover on the matter.
+
+byte[] refactored how yuzu emulates {{< gh-hovercard "12236" "CPU core activation," >}} he assures us writing three thousand lines of code simplifies the design of the ARM interface the emulator uses.
```
Comment (FearlessTobi) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:397]: I do not understand this sentence

**Review by FearlessTobi — COMMENTED** [review]
great stuff, I only did a very quick glance at the report though

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+
+As usual, [german77](https://github.com/german77) has been having his good share of fun working on input and the native applets.
+
+First in the list is a solution for Amiibos with read-only permissions. If a program mounted them as read-only, yuzu doesn’t access any of their encrypted data, and could mark them as corrupt, which wasn’t really the case.
+Since read-only data is set from the factory, {{< gh-hovercard "12255" "skipping the corruption check" >}} allows the use of read-only mounts in games like `The Legend of Zelda: Tears of the Kingdom`.
+
+Next up, an announcement. german77 started working on a rewrite for the HID (human interface devices) code, called `Project Leviathan` with the intention of further improving the accuracy of yuzu’s input emulation.
+So far only preliminary work has been done, but some results have already been put into service.
+
+The first part finished is the {{< gh-hovercard "12289" "emulation" >}} of `AppletResource`, which allows the developers to start working on multiprocess support in the near future, as well as other necessary resources like `AppletResourceUserId`, or simply aruid.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:377]:
```suggestion
The first part finished is the {{< gh-hovercard "12289" "emulation" >}} of `AppletResource`, which allows the developers to start working on multiprocess support in the near future, as well as other necessary resources like `AppletResourceUserId`, or simply `aruid`.
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+Sometimes, games have unexpected file organisation strategies.
+A recent episode of this phenomenon affects the `Batman: Arkham` trilogy: the base games ship with a null RomFS.
+That shouldn't have been a problem; having a RomFS is never a requirement to run a program. 
+`Batman: Arkham Asylum` did have a RomFS, and was launching fine, but `Batman: Arkham City` and `Batman: Arkham Knight` require their updates installed to be playable (their base game size is only 8MB).
+The update contains the RomFS, but since the base game didn't contain it, yuzu couldn't find the RomFS to load, even with the update installed.
+
+Handling this {{< gh-hovercard "12263" "peculiar case" >}} fell into byte[]’s hands, and the World’s Greatest Detective can now go fetch ? signs all over the place at ease.
+
+{{< imgs
+	"./batman.png|The Goddamn Batman (Batman: Arkham Knight)"
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:419]:
```suggestion
	"./batman.png|The Batman (Batman: Arkham Knight)"
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+  >}}
+
+Another {{< gh-hovercard "12392" "file system implementation" >}} byte[] gave us was support for the `OpenDirectoryMode` flag, which allows `Portal 2` to save and load properly. One has to wonder why this wasn’t fixed earlier.
+
+{{< imgs
+	"./portal21.png|Are we going to space? (Portal 2)"
+  >}}
+
+Some additional work was needed to get this masterpiece fully playable. 
+An entire service needed some patching-up, specifically, `ro`, or {{< gh-hovercard "12321" "relocatable object" >}} (allowing programs to load libraries on the fly), needed to be rewritten a bit–just over a thousand lines of code.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:429]:
```suggestion
An entire service needed some patching-up, specifically, `ro`, or {{< gh-hovercard "12321" "relocatable object" >}} (allowing programs to load libraries on the fly), needed to be rewritten a bit  — just over a thousand lines of code.
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+### UI changes
+
+german77 spotted an oopsie in the UI configuration files, leading to the language selection never getting saved.
+If anyone noticed yuzu always selected the default language, you now know what the reason was.
+{{< gh-hovercard "12402" "Making this value persistent" >}} solved the problem.
+
+And lastly one last minor buff to Linux desktop users, from the hands of newcomer [ReillyBrogan](https://github.com/ReillyBrogan).
+
+When using a Wayland compositor, program windows are matched to their `.desktop` shortcut file.
+The appId window property is intended to match the name of the `.desktop` file, for example, `org.yuzu_emu.yuzu.desktop`.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:455]:
```suggestion
The `appId` window property is intended to match the name of the `.desktop` file, for example, `org.yuzu_emu.yuzu.desktop`.
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+Plasma desktop by default sets this property to the name of the binary file, which in this case is just `yuzu`, and doesn’t match the expected value of `org.yuzu_emu.yuzu`.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:456]:
```suggestion
Plasma desktop, by default, sets this property to the name of the binary file, which in this case is just `yuzu`, and doesn’t match the expected value of `org.yuzu_emu.yuzu`.
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 9]
```diff
@@ -0,0 +1,539 @@
...
+This led to Plasma desktop users (Steam Deck included) displaying yuzu without our glorious icon.
+One {{< gh-hovercard "12521" "manual override" >}} later, and the icon is properly there. Thank you!
+
+## Hardware section
+
+Not much to mention this time, but there’s a few nits that deserve to be brought up.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:462]:
```suggestion
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 10]
```diff
@@ -0,0 +1,539 @@
...
+
+### Proprietary NVIDIA driver
+
+As previously mentioned, it’s safe to update to the latest driver versions thanks to the fixes implemented by byte[].
+This will become more critical once the SUPER refresh of Ada cards is out, so perfect timing there.
+Hopefully they finally provide good value.
+
+### NVK
+
+What’s this? A free and open source Mesa Vulkan driver for Turing and later NVIDIA hardware, that’s what it is!
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:474]:
```suggestion
What’s this? A free and open source Mesa Vulkan driver for Turing and later NVIDIA hardware?
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 11]
```diff
@@ -0,0 +1,539 @@
...
+We identified `Pokémon Scarlet & Violet` crashes on RDNA1 hardware exclusively (RX 5000 series).
+The usual standard procedure was followed, a custom test case was provided in a bug report.
+
+## Intel…
+
+In contrast, Intel failed us again. 
+No words from the fix for the geometry shader crash we [reported](https://github.com/IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT/issues/551) back in October yet.
+
+We hope it’s just a delay from the holidays.
+2024 is the year of ~~the Linux desktop~~ working Intel Windows drivers.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:490]:
```suggestion
In contrast, Intel failed us again. 
No word on the fix for the geometry shader crash we [reported](https://github.com/IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT/issues/551) back in October yet.

We hope it’s just a delay from the holidays.
2024 is the year of ~~the Linux desktop~~ working Intel Windows drivers.
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 12]
```diff
@@ -0,0 +1,539 @@
...
+
+## Qualcomm
+
+This section needs to be split too. Just a moment.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:494]:
```suggestion
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 13]
```diff
@@ -0,0 +1,539 @@
...
+
+### Proprietary Qualcomm driver
+
+Much better.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:498]:
```suggestion
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 14]
```diff
@@ -0,0 +1,539 @@
...
+
+We recently got our hands on a Snapdragon 8 Gen 3 device (a Red Magic 9), and while the official Qualcomm driver is still mediocre at best, we’re very impressed with the performance of the new Adreno 750 GPU. 
+In spite of its driver, it managed to bruteforce amazing framerate numbers. It still produced the same graphical glitches any other Adreno card has with their respective official drivers.
+
+Some games reached the 120 FPS limit imposed by the device’s screen refresh rate.
+In simpler terms, Balanced mode in the Red Magic 9 is faster than Diablo mode in the previous Red Magic 8.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:504]:
```suggestion
In short, Balanced mode in the Red Magic 9 is faster than Diablo mode in the previous Red Magic 8.
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 15]
```diff
@@ -0,0 +1,539 @@
...
+
+We can’t wait for Mesa to add Turnip support to this new hardware in town, but as anyone with experience using Mesa will tell you: "let them cook".
+
+Speaking of which.
+
+### Turnip
+
+Adreno 750 aside, which Mesa [already has in the oven](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/26934), the Turnip network of support continues to improve.
+Current [releases](https://github.com/K11MCH1/AdrenoToolsDrivers/releases) for example have added support for rare GPU variants, like Adreno 642L and Adreno 644.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:513]:
```suggestion
Current [releases](https://github.com/K11MCH1/AdrenoToolsDrivers/releases) have added support for rare GPU variants, like Adreno 642L and Adreno 644.
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 16]
```diff
@@ -0,0 +1,539 @@
...
+Performance has improved too, as well as stability.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:514]:
```suggestion
Performance and stability have improved as well.
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 17]
```diff
@@ -0,0 +1,539 @@
...
+
+Keep up to date with Turnip releases; they usually only bring benefits.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:516]:
```suggestion
Keep up to date with Turnip releases; they usually only bring improvements.
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 18]
```diff
@@ -0,0 +1,539 @@
...
+
+## Mali moment
+
+Total Mali moments count this month: 3.
+
+But it’s not all bad news dear Mali sufferers!
+
+[Newer generations](https://www.phoronix.com/news/Panthor-DRM-Newer-Mali) of Mali hardware, codenamed "Panthor" will have an officially backed Mesa driver, the already existing but almost abandoned [Panfrost/PanVK](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/26358) driver.
+This means in the future Mali moments will maybe stop existing, in a similar way to how Turnip provides a much better experience to Adreno users.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:525]:
```suggestion
This means that Mail moments may cease existing sometime in the future, in a similar way to how Turnip provides a much better experience to Adreno users.
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 19]
```diff
@@ -0,0 +1,539 @@
...
+
+## Future projects
+
+Multiprocess [GPU SMMU](https://github.com/yuzu-emu/yuzu/pull/12579) support is almost done now, thanks to Blinkhawk.
+We’ll cover it in more detail next month, but you can check the pull request draft meanwhile to see the expected benefits it brings.
+
+And maybe the rest of the team has something more going on, yeah. But let’s save that surprise for later, it’s more fun that way (this in no way means any attempt to leak a picture of the current progress was rejected by all devs, no no).
+
+That’s all folks! I’m finishing writing and collecting media at 7AM. Thank you for reading until the end, let’s see you all again next time in the first progress report of 2024!
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:534]:
```suggestion
That’s all folks! Thank you for reading until the end, we hope to see you all again in the first progress report of 2024!
```

**Review by Schplee — CHANGES_REQUESTED** [review]

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 20]
```diff
@@ -0,0 +1,539 @@
...
+Keep up to date with Turnip releases; they usually only bring improvements.
+
+## Mali moment
+
+Total Mali moments count this month: 3.
+
+But it’s not all bad news dear Mali sufferers!
+
+[Newer generations](https://www.phoronix.com/news/Panthor-DRM-Newer-Mali) of Mali hardware, codenamed "Panthor" will have an officially backed Mesa driver, the already existing but almost abandoned [Panfrost/PanVK](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/26358) driver.
+This means that Mail moments may cease existing sometime in the future, in a similar way to how Turnip provides a much better experience to Adreno users.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:525]:
```suggestion
This means that Mali moments may cease existing sometime in the future, in a similar way to how Turnip provides a much better experience to Adreno users.
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 21]
```diff
@@ -0,0 +1,539 @@
...
+Changing the logic to {{< gh-hovercard "12274" "ignore sRGB in framebuffer images" >}} provides smooth frametimes for users of these drivers.
+
+{{< imgs
+	"./sm3d.png|What does Mario do with that many stars? Not sing Peaches I hope (Super Mario 3D All-Stars)"
+  >}}
+
+On a related note, let’s talk about presentation limits and how they affect asynchronous presentation, which moves presenting to the screen to a different CPU thread.
+
+Due to several factors, yuzu’s Vulkan renderer could only process up to 6 frames at a time.
+While this isn’t normally an issue on desktop, especially for users that don’t enable Asynchronous Presentation (available in `Emulation > Configure… > Graphics > Advanced`), Android, a platform that demands always using asynchronous presentation, showed us that low enough performance with the setting toggled on, regardless of OS or platform, can cause the queue of swapchain images to grow beyond the limit of 6 frames. Leading to a driver crash and your progress being lost. Ouch.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:74]:
```suggestion
While this isn’t normally an issue on desktop, especially for users that don’t enable Asynchronous Presentation (available in `Emulation > Configure… > Graphics > Advanced`), Android, a platform that demands always using asynchronous presentation, showed us that low enough performance with the setting toggled on, regardless of OS or platform, can cause the queue of swapchain images to grow beyond the limit of 6 frames. Thus, leading to a driver crash and your progress being lost. Ouch.
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 22]
```diff
@@ -0,0 +1,539 @@
...
+### Proprietary Qualcomm driver
+
+
+We recently got our hands on a Snapdragon 8 Gen 3 device (a Red Magic 9), and while the official Qualcomm driver is still mediocre at best, we’re very impressed with the performance of the new Adreno 750 GPU. 
+In spite of its driver, it managed to bruteforce amazing framerate numbers. It still produced the same graphical glitches any other Adreno card has with their respective official drivers.
+
+Some games reached the 120 FPS limit imposed by the device’s screen refresh rate.
+In short, Balanced mode in the Red Magic 9 is faster than Diablo mode in the previous Red Magic 8.
+
+Even Turnip, with its [early support](https://github.com/K11MCH1/AdrenoToolsDrivers/releases/tag/v24.0.0_R15) already shows outstanding results, here is an Adreno 740 vs an Adreno 750:
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:500]:
```suggestion
Even Turnip, with its [early support](https://github.com/K11MCH1/AdrenoToolsDrivers/releases/tag/v24.0.0_R15) already shows outstanding results. Here is an Adreno 740 vs an Adreno 750:
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 23]
```diff
@@ -0,0 +1,539 @@
...
+
+{{< single-title-imgs-compare
+	"i9 14900K who? (The Legend of Zelda: Breath of the Wild)"
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:503]:
```suggestion
	"i9-14900K who? (The Legend of Zelda: Breath of the Wild)"
```

File: site/content/entry/yuzu-progress-report-dec-2023/index.md [hunk 24]
```diff
@@ -0,0 +1,539 @@
...
+Because [epicboy](https://github.com/ameerj) won’t let the old API rest.
+
+One remaining issue (of many) for the OG open API was a bug with the shadows of `Metroid Prime Remastered`.
+By bringing the {{< gh-hovercard "12412" "counter query accuracy," >}} which was first seen in Vulkan, to OpenGL, this issue is now solved.
+
+{{< imgs
+	"./metroid.png|SPAAAAACE (Metroid Prime Remastered)"
+  >}}
+
+Another win for OpenGL users is two fixes for `Xenoblade Chronicles 3`.
```
Comment (Schplee) [inline site/content/entry/yuzu-progress-report-dec-2023/index.md:157]:
```suggestion
OpenGL users saw another win with the implementation of two fixes for `Xenoblade Chronicles 3`.
```

**Review by Schplee — CHANGES_REQUESTED** [review]
